tests/MongoDB.Bson.Tests/Specifications/bson-corpus/BsonCorpusTestRunner.cs

```diff
@@ -0,0 +1,316 @@
+        private void RunDecodeErrorsTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "type", "description", "bson");
+
+            var bson = BsonUtils.ParseHexString(test["bson"].AsString);
+
+            var exception = Record.Exception(() =>
+            {
+                using (var stream = new MemoryStream(bson))
+                using (var reader = new BsonBinaryReader(stream))
+                {
+                    while (!reader.IsAtEndOfFile())
+                    {
+                        _ = BsonSerializer.Deserialize<BsonDocument>(reader);
+                    }
+                }
+            });
+
+            exception.Should().NotBeNull();
+        }
+
+        private void RunParseErrorsTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "type", "description", "string");
+
+            var json = test["string"].AsString;
+
+            var exception = Record.Exception(() => BsonDocument.Parse(json));
+
+            exception.Should().NotBeNull();
+        }
+
+        private void RunValidTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(
+                test,
+                "type",
+                "description",
+                "canonical_bson",
+                "canonical_extjson",
+                "relaxed_extjson",
+                "degenerate_bson",
+                "degenerate_extjson",
+                "converted_bson",
+                "converted_extjson",
+                "lossy");
+
+            byte[] cB = null;
+            if (test.Contains("canonical_bson"))
+            {
+                cB = BsonUtils.ParseHexString(test["canonical_bson"].AsString);
+            }
+
+            byte[] dB = null;
+            if (test.Contains("degenerate_bson"))
+            {
+                dB = BsonUtils.ParseHexString(test["degenerate_bson"].AsString);
+            }
+
+            string cEJ = null;
+            if (test.Contains("canonical_extjson"))
+            {
+                cEJ = test["canonical_extjson"].AsString.Replace(" ", "");
+                cEJ = UnescapeUnicodeCharacters(cEJ);
+            }
+
+            string dEJ = null;
+            if (test.Contains("degenerate_extjson"))
+            {
+                dEJ = test["degenerate_extjson"].AsString.Replace(" ", "");
+                dEJ = UnescapeUnicodeCharacters(dEJ);
+            }
+
+            string rEJ = null;
+            if (test.Contains("relaxed_extjson"))
+            {
+                rEJ = test["relaxed_extjson"].AsString.Replace(" ", "");
+                rEJ = UnescapeUnicodeCharacters(rEJ);
+            }
+
+            if (cB != null)
+            {
+                EncodeBson(DecodeBson(cB)).Should().Equal(cB, "native_to_bson( bson_to_native(cB) ) = cB");
+                EncodeCanonicalExtendedJson(DecodeBson(cB)).Should().Be(cEJ, "native_to_canonical_extended_json( bson_to_native(cB) ) = cEJ");
```

Comment:
Consider splitting this logic into several lines. I believe debugging of it will be not so simple

Reply:
This would lower readability while making debugging just a bit simpler. I'm leaving it as is, because in current state code perfectly correlates to specification, which makes it easy to track changes and understand the meaning of the test.
Consider specification description for this particular test scenario:
> for cB input:
> - native_to_bson( bson_to_native(cB) ) = cB
> - native_to_canonical_extended_json( bson_to_native(cB) ) = cEJ
> - native_to_relaxed_extended_json( bson_to_native(cB) ) = rEJ (if rEJ exists)

Not done.